python/ros_interface.py

```diff
@@ -0,0 +1,115 @@
+import rospy
+from sailbot_msg.msg import Sensors, windSensor, GPS
+
+# Constants
+CHECK_PERIOD_SECONDS = 0.1  # How often fields are updated
+KNOTS_TO_KMPH = 1.852
+ERROR = 0.2  # If signals need different margin of errors, modify as required
+
+"""
+Conversion Notes:
+- wind_speedKmph, wind_direction, wind_headingDegrees, gps_speedKmph: int32 -> float64 (no loss of precision)
+- wind_speedKmph, gps_speedKmph: knots -> km/h
+- wind_direction: (0 is forward, 90 is right, etc - cw - degrees) -> (0 is right, 90 is bow, etc - ccw - degrees)
+- heading_degrees: (0 is north, 90 is east, etc - cw - degrees) -> (0 is east, 90 is north, etc - ccw - degrees)
+"""
+
+
+class RosInterface:
+    def __init__(self):
+        rospy.init_node('ros_interface', anonymous=True)
+        rospy.Subscriber("sensors", Sensors, self.sensorsCallback)
+        self.pubWind = rospy.Publisher('windSensor', windSensor, queue_size=4)
+        self.pubGPS = rospy.Publisher('GPS', GPS, queue_size=4)
+        self.initialized = False
+        self.data = 0
+
+        self.wind_speedKmph = 0
+        self.wind_direction = 0
+        self.gps_lat = 0
+        self.gps_lon = 0
+        self.gps_headingDegrees = 0
+        self.gps_speedKmph = 0
+
+        self.past_wind_speedKmph = 0
+        self.past_wind_direction = 0
+        self.past_gps_lat = 0
+        self.past_gps_lon = 0
+        self.past_gps_headingDegrees = 0
+        self.past_gps_speedKmph = 0
+
+    def sensorsCallback(self, data):
+        if not self.initialized:
+            self.initialized = True
+        self.data = data
+
+    def pub(self):
+        if self.initialized:
+            self.translate()
+            self.pubGPS.publish(self.gps_lat, self.gps_lon, self.gps_headingDegrees, self.gps_speedKmph)
+            self.pubWind.publish(self.wind_direction, self.wind_speedKmph)
+        else:
+            rospy.logwarn("Tried to publish sensor values, but not initialized yet. Waiting for first sensor message.")
+
+    # Translate from many sensor outputs to one signal, checking error margin, averaging, and converting units
+    def translate(self):
+        data = self.data
+
+        # Past values
+        self.past_gps_lat = self.gps_lat
```

Comment:
Sorry one more thing. On first loop, these will all be 0's, which doesn't make sense. The error checking code will be messed up. You already have a check for the value to be 0, but doesn't work for the conversions (turns 0 to 90). 

Could init them all to `None` instead of 0, then have a None check in Err function